src/Modules/Legacy/Inverse/SolveInverseProblemWithTikhonovImpl.cc

```diff
@@ -188,103 +188,275 @@ namespace LinearAlgebra
+      //      M4 = identity
+      //      y = measuredData
     //.........................................................................
-    DenseMatrix ARRtrAtr, RRtrAtr;
-    double lambda=0, lambda_sq=0;
-    DenseMatrix forward_transpose = forwardMatrix_->transpose();
-    DenseMatrix regMat;
-    #ifdef SCIRUN4_CODE_TO_BE_CONVERTED_LATER
-    SparseRowMatrixHandle sourceWeighting_sparse;
-    #endif
-    DenseColumnMatrix solution(M);
-    DenseColumnMatrix RRtrAtrsolution(N);
+   
+    
+    DenseMatrix  RRtr, RRtrAtr, CCtr;
+    DenseMatrix iRRtr(N,N);
+    DenseMatrix iCCtr(M,M);
 
-    if (!sourceWeighting_) //check if a Source Space Weighting Matrix exist?
-    {
-      regMat = DenseMatrix::Identity(N, N);
-      ARRtrAtr = *forwardMatrix_ * forward_transpose;
-      RRtrAtr = forward_transpose;
-    }
-    else
-    {
-      if(((regularizationSolutionSubcase_==solution_constrained) && ((N != sourceWeighting_->nrows()) || (N != sourceWeighting_->ncols()))) ||
-         ((regularizationSolutionSubcase_==solution_constrained_squared) && (N != sourceWeighting_->nrows())))
-      {
-        BOOST_THROW_EXCEPTION(DimensionMismatch() << DimensionMismatchInfo("Solution Weighting Matrix (number of rows and columns) must fit matrix dimensions of Forward Matrix (number of columns) !"));
-      }
-      regMat = *sourceWeighting_;
-      #ifdef SCIRUN4_CODE_TO_BE_CONVERTED_LATER
-      sourceWeighting_sparse = sourceWeighting_->sparse();
-      #endif
+    // DEFINITIONS AND PREALOCATION OF SOURCE REGULARIZATION MATRIX 'R'
 
-      if (regularizationSolutionSubcase_== solution_constrained)
+      // if R does not exist, set as identity of size equal to N (columns of fwd matrix)
+      if (!sourceWeighting_)
       {
-        RRtrAtr = *sourceWeighting_ * forward_transpose;
-	      auto AR = forwardMatrix_;
-	      ARRtrAtr = *AR * RRtrAtr;
+          RRtr = DenseMatrix::Identity(N, N);
+          iRRtr = RRtr;
+          
       }
-      else if (regularizationSolutionSubcase_== solution_constrained_squared)
+      else
       {
-        auto AR = *forwardMatrix_ * *sourceWeighting_;
-	      auto RtrAtr = AR.transpose().eval();
-        RRtrAtr = *sourceWeighting_ * RtrAtr;
-	      ARRtrAtr = AR * RtrAtr;
+          
+        // if provided the non-squared version of R
+        if( regularizationSolutionSubcase_==solution_constrained )
+        {
+            
+            // check that the matrix is of appropriate size (equal number of rows as columns in fwd matrix)
+            if ( N != sourceWeighting_->nrows() )
+            {
+                BOOST_THROW_EXCEPTION(DimensionMismatch() << DimensionMismatchInfo("Solution Regularization Matrix must have the same number of rows as columns in the Forward Matrix !"));
+            }
+          
+            RRtr = *sourceWeighting_ * sourceWeighting_->transpose();
+            iRRtr = RRtr.inverse();   // todo: @JCOLLFONT need to compute inverse of RRtr and set try/catch
```

Comment:
@jcollfont The fix for this and similar compile errors is to force evaluation. Eigen likes to optimize assignments and in this case doesn't want to compute the inverse unless you tell it explicitly:
```
iRRtr = RRtr.inverse().eval();
```
That's it. Same for other inverse calls.